src/Tools/ExternalAccess/LiveShare/Tagger/SyntacticClassificationModeSelector.cs

```diff
@@ -0,0 +1,71 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Threading;
+using Microsoft.CodeAnalysis.Experiments;
+using Microsoft.CodeAnalysis.ExternalAccess.LiveShare.Classification;
+using Microsoft.CodeAnalysis.Text;
+using Microsoft.VisualStudio.Text;
+
+namespace Microsoft.CodeAnalysis.ExternalAccess.LiveShare.Tagger
+{
+    /// <summary>
+    /// this is almost straight copy from typescript for syntatic LSP experiement.
+    /// we won't attemp to change code to follow Roslyn style until we have result of the experiement
+    /// </summary>
+    internal sealed class SyntacticClassificationModeSelector
+    {
+        private readonly AbstractLspClientServiceFactory _lspClientServiceFactory;
+        private readonly ITextBuffer _buffer;
+
+        private int _modeCache = -1;
+
+        public SyntacticClassificationModeSelector(AbstractLspClientServiceFactory lspClientServiceFactory, ITextBuffer buffer)
+        {
+            _lspClientServiceFactory = lspClientServiceFactory;
+            _buffer = buffer;
+        }
+
+        public static SyntacticClassificationModeSelector GetModeSelector(AbstractLspClientServiceFactory lspClientServiceFactory, ITextBuffer buffer)
+        {
+            return new SyntacticClassificationModeSelector(lspClientServiceFactory, buffer);
+        }
+
+        public SyntacticClassificationMode GetMode()
```

Comment:
>GetMode [](start = 43, length = 7)

TS used a more complicated event driven model because we called this in very hot code paths and it could change at runtime.  If yours can't change at runtime, a simple cache is probably sufficient.

Reply:
Can you clarify what "change at runtime" means?

Reply:
Our decision logic depends on things like the TS version that the user can change during a VS session.

Reply:
Sure.  But that wouldn't be something difficult for Roslyn to do.  We have tons of features that need to behave differently based on things like versions (or other ambient things that could change).  Consider, for example, `#if` preprocessor directives.  If those ambient project/env values change, then we need to redo everything as well.

Anyways, i'm not going to belabor hte point.  If this is how things are written, it's how things are written.  It was just surprising to me given how perf tuned Roslyn is for this scenario already.  in general, i wish (if we ran into perf issues) we'd target the fixes inside roslyn so all the langauges could benefit instead of having a specific langauge fork off and reimplement everything.

Reply:
AFAIK, this type originated in TS, rather than in Roslyn (i.e. I wrote it and I don't remember following an existing example).  Since it seems to have been simplified during the port, I thought I'd give some context for why there was additional complexity in the original.